frontend/src/lib/rest/RestRequests.ts

```diff
@@ -0,0 +1,83 @@
+import { showNotification } from 'modules/notifications/notifications';
+
+const username = 'admin';
+const password = 'password';
+const backendUrl = process.env.REACT_APP_BACKEND_HOST;
+const DEFAULT_TIMEOUT = 10000;
+
+export enum Method {
+  GET = 'GET',
+  POST = 'POST',
+  PUT = 'PUT',
+  DELETE = 'DELETE',
+}
+
+const getHeaders = (): Headers => {
+  return new Headers({
+    Authorization: `Basic ${btoa(`${username}:${password}`)})`,
+    'Content-Type': 'application/json',
+  });
+};
+
+const timeout = (promise: Promise<Response>, ms = DEFAULT_TIMEOUT): Promise<Response> => {
+  const timerPromise = new Promise<never>((_, reject) => {
+    setTimeout(() => {
+      reject(new Error('Timeout'));
+    }, ms);
+  });
+  return Promise.race([timerPromise, promise]);
+};
+
+const call = <T>(method: Method, path: string, data: unknown): Promise<T | void> => {
+  const url = `${backendUrl}${path}`;
+  const options = {
+    method,
+    headers: getHeaders(),
+    body: data ? JSON.stringify(data) : null,
+  };
+  return timeout(fetch(url, options)).then(response => {
```

Comment:
We could make these methods use async syntax to get rid of the promise chains. For example

```
const call = async <T>(method: Method, path: string, data: unknown): T | void => {
    const url = `${backendUrl}${path}`;
    const options = {
        method,
        headers: getHeaders(),
        body: data ? JSON.stringify(data) : null
    };

    const response = await timeout(fetch(url, options));
    if (!response.ok) {
        throw new Error(`${response.status} ${response.body}`);
    }

    if (response.status === 204) {
        return null;
    }

    const text = await response.text();

    return text ? JSON.parse(text) : response;
};
```

Thoughts?

Reply:
I had the same thought, and yeah I think it could be better, but I was holding off on that for now just because I think these things can be simplified further as we migrate the other calls over. Currently there are extra cases in here just to support some components which are doing weird things - eg the last line is returning the whole response object rather than just the data because some callers are expecting that.

Reply:
Yeah thats cool I will merge it in then and we can revisit this once we get a little closer